File: ProductOfArrayExceptSelf.cs
// Time Complexity : O(n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode : yes
// Three line explanation of solution in plain english
// using running product, calculate left product and store it in result
// and in reverse loop calculate right product and multiply it with result(which has the left product)
public class Solution {
    public int[] ProductExceptSelf(int[] nums) {
        int n = nums.Length;
        int[] result = new int[n];
        int rp = 1;
        result[0] = rp;
        //left pass
        for(int i=1; i<n;i++){
            rp = nums[i-1]*rp;
            result[i] = rp;
        }

        rp=1;
        //right pass
        for(int i=n-2; i>=0; i--){
            rp = nums[i+1] * rp;
            result[i] = result[i] * rp; //already available result which contains left product * right product
        }
         return result;
    }
}

File: Product of Array Except Self.java
// TC = O(n)
// SC = O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no

class Solution {
    public int[] productExceptSelf(int[] nums) {
        // maintain running product till that index from L & R and then multiply it at index. eg. result[i] = left[i] * right[i]
        int n = nums.length;
        int[] result = new int[n];
        // null
        if(nums == null || n == 0) return result;
        int pass = 1;
        result[0] = 1; // initialize result[0] with 1
        // left pass
        for(int i = 1; i < n; i++) { 
            pass *= nums[i - 1];
            result[i] = pass;
        }
        // right pass
        pass = 1; // reset pass
        for(int i = n - 2; i >= 0; i--) {
            pass *= nums[i + 1];
            result[i] *= pass; // IMP multiply result with pass to get left and right product
        }
        return result;
    }
}